Inc/HALAL/Models/Packets/Packet.hpp

```diff
@@ -0,0 +1,190 @@
+#pragma once
+
+#include "PacketValue.hpp"
+
+struct empty_type{};
+
+template<class... ValueTypes> class Packet;
+
+template<>
+class Packet<>
+{
+public:
+	uint16_t id;
+	using value_type = empty_type;
+	static constexpr size_t number_of_values = 0;
+	using base_type = empty_type;
+	static map<decltype(Packet<>::id), function<void(uint8_t*)>> save_by_id;
+	static map<decltype(Packet<>::id), void(*)()> on_received;
+
+public:
+	Packet(int id): id(static_cast<decltype(this->id)>(id)){}
+	Packet() = default;
+
+	static decltype(id) get_id(uint8_t* new_data){
+		return *(decltype(id)*)new_data;
+	}
+};
+
+template<class Type, class... Types>
+class Packet<Type, Types...> : public Packet<Types...>
+{
+public:
+	uint16_t id = 0;
+	uint8_t* buffer = nullptr;
+	using value_type = Type;
+	static constexpr size_t number_of_values = 1 + sizeof...(Types);
+	using base_type = Packet<Types...>;
+
+protected:
+	PacketValue<value_type> value;
+
+public:
+	Packet();
+	Packet(int id);
+	Packet(int id, PacketValue<Type> value, PacketValue<Types>... args);
+
+	template<int I>
+	const auto& get() const;
+
+	template<int I>
+	auto& get();
+
+	uint8_t* build();
+
+	template <size_t I = 0>
+	void calculate_sizes();
+
+	template <size_t I = 0>
+	void fill_buffer();
+
+	bool check_id(uint8_t* new_data);
+
+    decltype(Packet<Type,Types...>::id) get_id();
+
+    template <size_t I = 0>
+    void load_data(uint8_t* new_data);
+
+    void save_data(uint8_t* new_data);
+
+public:
+	uint16_t buffer_size = sizeof(id);
+	uint16_t ptr_loc = sizeof(id);
+	bool has_been_built = false;
+};
+
+#if __cpp_deduction_guides >= 201606
```

Comment:
```suggestion
#if __cpp_deduction_guides >= 201606
```
Para que es este bloque de codigo?

Reply:
el compilador que usamos es extraño y no entiende las deduction guides bien con eso no se porque pero ya no se raya y compila, si lo quitas no sabe hacer la inferencia de las templates, es copiado de una clase de la std lib de c++